File: WorldFeed/Server/src/Services/Upload/WorldFeed.Upload.API/Client/Clients/IExecuteClient.cs
namespace WorldFeed.Upload.Client.Clients
{
    using System;
    using System.Threading.Tasks;

    using WorldFeed.Upload.API.Web;
    using WorldFeed.Upload.Application.Query;
    using WorldFeed.Upload.Application.Requesters;

    public interface IExecuteClient
    {
        /// <summary>
        /// Execute a custom request
        /// </summary>
        /// <returns>The raw response from twitter with the json parsed into a Data Transfer Object</returns>
        Task<ITwitterResult<T>> AdvanceRequestAsync<T>(Action<ITwitterRequest> configureRequest) where T : class;

        /// <summary>
        /// Execute a custom request
        /// </summary>
        /// <returns>The raw response from Twitter</returns>
        Task<ITwitterResult> AdvanceRequestAsync(Action<ITwitterRequest> configureRequest);

        /// <summary>
        /// Execute a custom query
        /// </summary>
        /// <returns>The raw response from twitter with the json parsed into a Data Transfer Object</returns>
        Task<ITwitterResult<T>> RequestAsync<T>(Action<ITwitterQuery> configureQuery) where T : class;

        /// <summary>
        /// Execute a custom query
        /// </summary>
        /// <returns>The raw response from Twitter</returns>
        Task<ITwitterResult> RequestAsync(Action<ITwitterQuery> configureQuery);
    }
}

File: Sharebook/Client/src/app/core/Public/Client/Clients/IExecuteClient.ts
import {Inject, InjectionToken} from "@angular/core";

import {ITwitterRequest} from "../../Models/Interfaces/ITwitterRequest";
import {ITwitterResult} from "../../../Core/Web/TwitterResult";
import {ITwitterQuery} from "../../Models/Interfaces/ITwitterQuery";
import {ExecuteClient} from "../../../../sharebook/Client/Clients/ExecuteClient";
import {ExecuteRequester} from "../../../../sharebook/Client/Requesters/ExecuteRequester";

export interface IExecuteClient {
  // Execute a custom request
  // <returns>The raw response from twitter with the json parsed into a Data Transfer Object</returns>
  advanceRequestAsync<T>(configureRequest: (twitterRequest: ITwitterRequest) => void): Promise<ITwitterResult<T>>;

  // Execute a custom request
  // <returns>The raw response from Twitter</returns>
  advanceRequestAsync(configureRequest: (twitterRequest: ITwitterRequest) => void): Promise<ITwitterResult>;

  // Execute a custom query
  // <returns>The raw response from twitter with the json parsed into a Data Transfer Object</returns>
  requestAsync<T>(configureQuery: (twitterQuery: ITwitterQuery) => void): Promise<ITwitterResult<T>>;

  // Execute a custom query
  // <returns>The raw response from Twitter</returns>
  requestAsync(configureQuery: (twitterQuery: ITwitterQuery) => void): Promise<ITwitterResult>;
}

export const IExecuteClientToken = new InjectionToken<IExecuteClient>('IExecuteClient', {
  providedIn: 'root',
  factory: () => new ExecuteClient(Inject(ExecuteRequester)),
});
